sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,448 @@
+            Squared euclidean norm of each data point. Pass it if you have it
+            at hands already to avoid it being recomputed here.
+
+        init : {'k-means++', 'random'}, callable or ndarray of shape \
+                (n_clusters, n_features)
+            Method for initialization.
+
+        random_state : RandomState instance
+            Determines random number generation for centroid initialization.
+            See :term:`Glossary <random_state>`.
+
+        init_size : int, default=None
+            Number of samples to randomly sample for speeding up the
+            initialization (sometimes at the expense of accuracy).
+        Returns
+        -------
+        centers : ndarray of shape (n_clusters = 2, n_features)
+        """
+        n_samples = X.shape[0]
+        n_clusters = 2
+
+        if init_size is not None and init_size < n_samples:
+            init_indices = random_state.randint(0, n_samples, init_size)
+            X = X[init_indices]
+            x_squared_norms = x_squared_norms[init_indices]
+            n_samples = X.shape[0]
+
+        if isinstance(init, str) and init == 'k-means++':
+            centers, _ = _kmeans_plusplus(X, n_clusters,
+                                          x_squared_norms, random_state)
+        elif isinstance(init, str) and init == 'random':
+            seeds = random_state.permutation(n_samples)[:n_clusters]
+            centers = X[seeds]
+        elif hasattr(init, '__array__'):
+            centers = init
+        elif callable(init):
+            centers = init(X, n_clusters, random_state=random_state)
+            centers = check_array(centers,
+                                  dtype=X.dtype, copy=False, order='C')
+            self._validate_center_shape(X, centers)
+
+        if sp.issparse(centers):
+            centers = centers.toarray()
+
+        return centers
+
+    def _bisect(self, X, y=None, sample_weight=None, random_state=None):
+        """ Bisection of data
+        Attempts to get best bisection of data by performing regular K-Means
+        for different pairs of centroids
+
+        .. note:: Number of attempts is specified by self.n_init
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+
+            Training instances to cluster.
+
+            .. note:: The data will be converted to C ordering,
+                which will cause a memory copy
+                if the given data is not C-contiguous.
+
+        y : Ignored
+            Not used, present here for API consistency by convention.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            The weights for each observation in X. If None, all observations
+            are assigned equal weight.
+
+        Returns
+        -------
+        self
+            Fitted estimator.
+        """
+        X = self._validate_data(X, accept_sparse='csr',
```

Comment:
This validation happening at each bisecting iteration seems inefficient. Is it necessary ?

Reply:
I'm not also satisfied from that part, but for now I don't have better idea

Reply:
> Removing that part would cause estimator to produce wrong centroids.

Can you explain why it would be wrong ?

Reply:
My fault - At bisecting X is substacted by mean but not restored after operation.  I'll remove that and unnecassary mean subtraction